deps/ustring.lua

```diff
@@ -0,0 +1,188 @@
+local ustring = {}
+local tostring = tostring
+local _meta = {}
+
+local strsub = string.sub
+local strbyte = string.byte
+local band = bit.band
+local rshift = bit.rshift
+function ustring.new(str,allowInvaild)
+    local str = str and tostring(str) or ""
+    local ustr = {}
+    local index = 1;
+    local append = 0;
+    for i = 1,#str do
+        repeat
+        local char = strsub(str,i,i)
+        local byte = strbyte(char)
+        if append ~= 0 then
+            if not allowInvaild then
+                if rshift(byte,6) ~= 0x02 then
+                    error("Invaild UTF-8 sequence at "..i)
+                end
+            end
+            ustr[index] = ustr[index] .. char
+            append = append - 1
+            if append == 0 then
+                index = index + 1
+            end
+            break
+        end
+
+        if rshift(byte,7) == 0x00 then -- 1 byte (ansi)
+            ustr[index] = char
+            index = index + 1
+        elseif rshift(byte,5) == 0x06 then -- 2 bytes
+            ustr[index] = char
+            append = 1
+        elseif rshift(byte,4) == 0x0E then -- 3 bytes
+            ustr[index] = char
+            append = 2
+        elseif rshift(byte,3) == 0x1E then -- 4 bytes
+            ustr[index] = char
+            append = 3
+        else
+            -- RFC 3629 (2003.11) says UTF-8 don't have characters larger than 4 bytes.
+            -- They will not be processed although they may be appeared in some old systems.
+            if not allowInvaild then
+                error("Invaild UTF-8 sequence at "..i..",byte:"..byte)
+            end
+        end
+
+        until true
+    end
+    setmetatable(ustr,_meta)
+    return ustr
+end
+
+function ustring.copy(ustr)
+    local u = ustring.new()
+    for i = 1,#ustr do
+        u[i] = ustr[i]
+    end
+    return u
+end
+
+function ustring.uindex(ustr,rawindex,initrawindex,initindex)
+    -- get the index of the UTF-8 character from a raw string index
+    -- return `nil` when rawindex is invaild
+    -- the last 2 arguments is used for speed up
+    local index = (initindex or 1)
+    rawindex = rawindex - (initrawindex or 1) + 1
+    repeat
+        local byte = ustr[index]
+        if byte == nil then return nil end
+        local len = #byte
+        index = index + 1
+        rawindex = rawindex - len
+    until rawindex <= 0
+    return index - 1
+end
+
+local gsub = string.gsub
+local find = string.find
+local format = string.format
+local gmatch = string.gmatch
+local match = string.match
+local lower = string.lower
+local upper = string.upper
+ustring.len = rawlen
```

Comment:
`rawlen`is lua's builtin function.
https://www.lua.org/manual/5.2/manual.html#pdf-rawlen